test/models/user_test.exs

```diff
@@ -0,0 +1,86 @@
+defmodule Yggdrasil.UserTest do
+  use Yggdrasil.ModelCase
+
+  alias Yggdrasil.User
+
+  @valid_attrs %{username: "tester", password: "password", password_confirmation: "password"}
+  @valid_attrs_upcase_user %{username: "TESTER", password: "password", password_confirmation: "password"}
+
+  @missing_username %{password: "password", password_confirmation: "pass"}
+  @missing_password %{username: "tester", password_confirmation: "password"}
+  @missing_password_conf %{username: "tester", password: "password"}
+  @password_mismatch %{username: "tester", password: "test", password_confirmation: "tester"}
+  @password_bad_len %{username: "tester", password: "te", password_confirmation: "tester"}
+  @password_conf_bad_len %{username: "tester", password: "te", password_confirmation: "te"}
+
+  @blank_msg "can't be blank"
+  @doesntmatch_msg "does not match confirmation"
+  @unique_msg "has already been taken"
+
+  @invalid_length_msg "should be at least %{count} character(s)"
+  @min_count 4
+
+  test "changeset with valid attributes" do
+    changeset = User.changeset(%User{}, @valid_attrs)
+    assert changeset.valid?
+  end
+
+  test "changeset with missing username" do
+    changeset = User.changeset(%User{}, @missing_username)
+
+    assert changeset.errors[:username] == @blank_msg
+    refute changeset.valid?
+  end
+
+  test "changeset with missing password" do
+    changeset = User.changeset(%User{}, @missing_password)
+
+    assert changeset.errors[:password] == @blank_msg
+    refute changeset.valid?
```

Comment:
http://hexdocs.pm/ecto/Ecto.Changeset.html#cast/4

ha, don't even have to check so basically if create_update_password only requires password/password_confirmation even if username is provided it will be ignored. It may error not sure need to test, but I think it ignores them. So no need to explicitly check that it was not provided which is nice.

I can do this for update_changeset as well where I don't want the hash changed, well actually now I see why it was not required in the changeset, make senses. They didn't require it because it should only be generated after validations happen and not provided in this case.

I think I have a handle on this now. going to submit some changes in a branch and let you review. If you are cool with that.